src/launcher/java/org/truffleruby/launcher/RubyLauncher.java

```diff
@@ -53,8 +55,27 @@ public static void main(String[] args) {
         new RubyLauncher().launch(args);
     }
 
+    public static boolean isGraalVM() {
+        return System.getProperty("org.graalvm.home") != null;
+    }
+
     // TODO (pitr-ch 22-Feb-2018): replace with a call to sdk API when available
     static boolean isGraal() {
+        if (isGraalVM() || isAOT()) {
+            return true;
+        }
+
+        try {
+            final Class<?> truffleClass = RubyLauncher.class.getClassLoader().loadClass("com.oracle.truffle.api.Truffle");
+            final Class<?> truffleRuntimeClass = RubyLauncher.class.getClassLoader().loadClass("com.oracle.truffle.api.TruffleRuntime");
+            final Method getTruffleRuntime = truffleClass.getDeclaredMethod("getRuntime");
+            final Method getName = truffleRuntimeClass.getDeclaredMethod("getName");
+
+            return ((String) getName.invoke(getTruffleRuntime.invoke(null))).toLowerCase(Locale.ENGLISH).contains("graal");
+        } catch (NoSuchMethodException | ClassNotFoundException | InvocationTargetException | IllegalAccessException e) {
+            // ignore and fallback to using Ruby Context
```

Comment:
Is there a case where this fails in practice?

Reply:
Probably not, but since the whole thing will be eventually replaced I don't want to spend time on figuring out what might go wrong with the reflection.

Reply:
The fact it will eventually be replaced is not a reason to keep big hacks around.
As I said above, please remove the "creating an extra RubyContext to get a boolean out" hack. And then just let reflection fail as it's a bug of `isGraal()` if it fails.
Nothing to figure out, just wrap those reflection exceptions in a RuntimeException.

Reply:
The reflection solution fails in CI. Christian already started working on adding the API so I'll wait for it instead of spending time on these hacks.